recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl

```diff
@@ -371,13 +369,13 @@ class EnvFirmware(Firmware):
                     firmware_size = os.path.getsize(env_default_binary.name)
                     self.firmware = open(env_default_binary.name, "rb")
 
-                    while True:
-                        if firmware_size <= 0:
-                            break
-                        firmware_size = self.mtd_device.write(
-                            mtd_dev_path, mtd_size, mtd_erasesize,
-                            env_default_binary, firmware_size
-                        )
+                    firmware_size = self.mtd_device.write(
+                        mtd_dev_path, mtd_size, mtd_erasesize,
+                        env_default_binary, firmware_size
+                    )
+
+                    if firmware_size > 0:
+                        raise UpgradeError("Write env failed")
```

Comment:
Are you sure it can write all the bytes in one shot? I think the original logic is trying to avoid if the data could not be written in just one write.

Reply:
If we keep the loop, there will be a problem; If somehow it always fails to write, a dead loop will happen here.

On the high level there are 3 times retries to improve this situation.

Reply:
Also there is a loop in `mtd_device.write` to confirm all data are touched.

```py
            while mtd_pos < mtd_size and file_size > 0:
                mtd_content = os.read(mtd_dev, mtd_erasesize)
                firmware_content = file_obj.read(mtd_erasesize)
                padsize = mtd_erasesize - len(firmware_content)
                firmware_content += bytearray([0xff] * padsize)

                if not mtd_content == firmware_content:
                    self.__erase(mtd_dev, mtd_pos, mtd_erasesize)
                    os.lseek(mtd_dev, mtd_pos, os.SEEK_SET)
                    os.write(mtd_dev, firmware_content)

                mtd_pos += mtd_erasesize
                file_size -= mtd_erasesize
```